### PR TITLE
Ruby enumerable methods project

### PR DESCRIPTION
**Pull request by Salvador-ON**
In this Ruby project we recreate the enumerable methods from scratch.

in my readme file is the live version in repl.it of my code

For the educational purpose in the steps that the Odin project request to modify the methods, I created another one, for example, my_map, my_map_2, my_map3

In the last part of the code are all the tests that I used to test my code

### CODE REVIEW

**Review by fatymahmed — CHANGES_REQUESTED** [review]
## Hi @Salvador-ON 
Great job so far 👏👏
Screenshot
As you can see from the image above, the common error is `no block given (yield)`
- If you have yield within your method definition, that means you have to obligatorily pass a block when you use it, that's why it's failing for the tests where no block is passed.
- For example for the `each_with_index` method, without passing a block, it returns an Enumerator.
`[2,3,4,5].each_with_index`
returns 
`#<Enumerator: [2, 3, 4, 5]:each_with_index>`
For your `my_each_with_index` method, when you run it without a block (just like the one above), it produces an error `no block given (yield) `, that is because it expects a block.
- Go through all your methods and make sure they cover both cases where a block is passed as an argument and the other case is when there's no block.
- Check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) to know the different arguments that can be passed to these methods so you can have all cases.

### Kindly make the suggested changes then request for another code review.
### Status: Changes Required 🛑
_Feel free to reach out to me on slack (Fatima) if you have any questions, I'd be happy to help._

**Comment by Salvador-ON** [comment]
I fix all the reported issue also I only left the final my_map

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+Style/FrozenStringLiteralComment:
+  EnforcedStyle: never
+Style/CaseEquality:
+  Enabled: false
+Metrics/PerceivedComplexity:
+  Enabled: false
+Metrics/CyclomaticComplexity:
+  Enabled: false
+Metrics/ModuleLength:
+  Enabled: false
```
Comment (bolah2009) [inline .rubocop.yml]: [Optional]

Please let's use the [**correct config files**](https://github.com/microverseinc/linters-config/tree/master/ruby).


> Do not make any changes in config files - they represent style guidelines that you share with your team - which is a group of all Microverse students. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/))


You can disable some rules when they are absolutely necessary in the source code. See below:

> `Style/CaseEquality` cop can be disabled as described here https://rubydoc.info/gems/rubocop/0.24.1/frames#Disabling_Cops_within_Source_Code ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/style_case_equality.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421589/))

You can also open PR in the linter config repo with a detailed description if you think it is necessary to change the current config


#### _Note that this is a required change but marked as optional as @fatymahmed didn't mention it in the first review._

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+arr5 = [3, 3, nil, 3]
+arr6 = [3, 3, 3]
+arr7 = []
+x2 = proc { |num| num * 2 }
+
+puts '-----my each-----'
+p(arr.my_each { |num1| num1 })
+puts '-----my each with index-----'
+hash = {}
+%w[cat dog wombat].my_each_with_index { |item, index| hash[item] = index }
```
Comment (bolah2009) [inline main.rb]: [Optional]

Let's remove all this code (and all other occurrences). 

We can, however, place it in a different file for the purpose of testing the methods. Also, we can add a few of this to the readme to illustrate the functions of the methods written. 

In future project when you will have to write a test or reuse these methods in another file, this example will create unnecessary noise.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      end
+    else
+      status = true
+      iarr.my_each do |num|
+        status = true if num && num != false && status == true
+        status = false if (num.nil? || num == false) && status == true
+      end
+    end
+    status
+  end
```
Comment (bolah2009) [inline main.rb]: [Optional]

for `my_all`, `my_none` and `my_any` let's handle cases when the argument is a `REGEX` or a `CLASS`. See example below (using the example on ruby doc https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-all-3F):


```ruby
# Using the built-in method
irb(main):010:0> %w[ant bear cat].all?(/a/)
#=> true

# Using my_all? method
irb(main):011:0> %w[ant bear cat].my_all?(/a/)
# => false

# Using the built-in method
irb(main):012:0> [1, 2i, 3.14].all?(Numeric) 
#=> true

# Using my_all? method
irb(main):013:0> [1, 2i, 3.14].my_all?(Numeric) 
#=> false
```

Hint: Use ` === ` to check if a member matches a patern.

#### _Note that this is a required change but marked as optional as @fatymahmed didn't mention it in the first review._

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        status = true if num == val && status == false
+      end
+    else
+      status = false
+      iarr.my_each do |num|
+        status = true if num && status == false
+      end
+    end
+    status
+  end
```
Comment (bolah2009) [inline main.rb]: [Optional]

See the previous comment on handling cases where the argument is a `Regex` or a `Class`.

### Find more details about [`any?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      iarr.my_each do |num|
+        status = false if val == num
+      end
+    elsif val.nil? && !block_given?
+      iarr.my_each do |num|
+        status = false if num
+      end
+    end
+    status
+  end
```
Comment (bolah2009) [inline main.rb]: [Optional]

See the previous comment on handling cases where the argument is a `Regex` or a `Class`.

### Find more details about [`none?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      yield(iarr[x])
+      x += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum :my_each_with_index unless block_given?
+
+    iarr = self
```
Comment (bolah2009) [inline main.rb]: [Optional]


As you have done in `my_each`, let's use `to_a` instead, This helps handle cases where the enumerable is a `Range`.
Currently using a `Range`, `(0..5).my_each_with_index{|a,b| puts "#{a} #{b}"}` gives an error.

**Review by bolah2009 — APPROVED** [review]
It's a 👍🏾 from me!

### :heavy_check_mark: Approved!

This project is approved, however, I would recommend checking out the optional comments before merging. :relaxed: 

Since the objective is to create identical methods to the ruby's enumerable methods, I would recommend resolving the comments I left below.

As said in the comment, I marked many of them as optional as @fatymahmed didn't explicitly ask you to make changes with them even though @fatymahmed requested to check the ruby doc.
> Check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) to know the different arguments that can be passed to these methods so you can have all cases.


Good luck with your next projects. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Comment by Salvador-ON** [comment]
ok thank you very much for the feedback, I will make the changes

**Pull request merged**